core/src/cli_commands/update_pool.rs

```diff
@@ -0,0 +1,345 @@
+        proptest::arbitrary::any_with::<TokenId>(
+            ergo_lib::ergotree_ir::chain::token::arbitrary::ArbTokenIdParam::Arbitrary,
+        )
+        .new_tree(&mut Default::default())
+        .unwrap()
+        .current()
+    }
+    #[test]
+    fn test_update_pool_box() {
+        let ctx = force_any_val::<ErgoStateContext>();
+        let height = ctx.pre_header.height;
+
+        let ballot_token_id = force_any_tokenid();
+        let pool_nft_token_id = force_any_tokenid();
+        let update_nft_token_id = force_any_tokenid();
+        let reward_token_id = force_any_tokenid();
+        let reward_tokens = Token {
+            token_id: reward_token_id,
+            amount: 1500.try_into().unwrap(),
+        };
+        let update_contract = UpdateContract::new()
+            .with_min_votes(5)
+            .with_pool_nft_token_id(pool_nft_token_id.clone())
+            .with_ballot_token_id(ballot_token_id.clone());
+        let mut update_box_candidate = ErgoBoxCandidateBuilder::new(
+            BoxValue::SAFE_USER_MIN,
+            update_contract.ergo_tree(),
+            height,
+        );
+        update_box_candidate.add_token(Token {
+            token_id: update_nft_token_id.clone(),
+            amount: 1.try_into().unwrap(),
+        });
+        let update_box = ErgoBox::from_box_candidate(
+            &update_box_candidate.build().unwrap(),
+            force_any_val::<TxId>(),
+            0,
+        )
+        .unwrap();
+
+        let pool_contract = PoolContract::new()
+            .with_refresh_nft_token_id(force_any_tokenid())
+            .with_update_nft_token_id(update_nft_token_id.clone());
+        let pool_box_candidate = make_pool_box_candidate(
+            &pool_contract,
+            0,
+            0,
+            Token {
+                token_id: pool_nft_token_id.clone(),
+                amount: 1.try_into().unwrap(),
+            },
+            reward_tokens.clone(),
+            BoxValue::SAFE_USER_MIN,
+            height,
+        )
+        .unwrap();
+        let pool_box =
+            ErgoBox::from_box_candidate(&pool_box_candidate, force_any_val::<TxId>(), 0).unwrap();
+
+        let new_refresh_token_id = force_any_tokenid();
+        let new_pool_contract = PoolContract::new()
+            .with_refresh_nft_token_id(new_refresh_token_id)
+            .with_update_nft_token_id(update_nft_token_id.clone());
+
+        let pool_box_bytes = new_pool_contract
+            .ergo_tree()
+            .sigma_serialize_bytes()
+            .unwrap();
+        let pool_box_hash = blake2b256_hash(&pool_box_bytes);
+
+        let ballot_contract = BallotContract::new()
+            .with_min_storage_rent(*BoxValue::SAFE_USER_MIN.as_u64())
+            .with_update_nft_token_id(update_nft_token_id);
+        let mut ballot_boxes = vec![];
+
+        let secret = force_any_val::<DlogProverInput>();
+        for _ in 0..5 {
+            //let secret = force_any_val::<DlogProverInput>();
```

Comment:
This test fails if I uncomment this line, which means each ballot box has the same public key as the wallet. The error is thread 

>'cli_commands::update_pool::tests::test_update_pool_box' panicked at 'called `Result::unwrap()` on an `Err` value: TxSigningError(ProverError(TreeRootIsNotReal, 2))', core/src/cli_commands/update_pool.rs:343:57

I'm guessing that when signing the transaction, it also tries to sign all the ballot boxes with your pubkey, and thus proveDlog fails? I'm not quite sure how to avoid this. Second, it seems you need a TransactionContext, and you have to include all the input boxes in that, which is a bit tedious. Should build_update_pool_box return a TransactionContext instead of UnsignedTransaction?

Reply:
Hmm, I tried that and I am running into the same issue. I believe the reason is in the ballot contract, where 
```
  val owner = proveDlog(selfPubKey)
  
  // unlike in collection, here we don't require spender to be one of the ballot token holders
  isSimpleCopy && (owner || update)
```

Since we're signing the transaction, the signer also tries to sign the ballot tokens, even though we don't need to be the owner to spend them in this case. But it does so anyways, and then proveDlog fails, which is why having the same pubkey for all ballots and the wallet works, but having different ones isn't. Is there any way to tell the signer that we don't need to sign the ballot boxes to spend them?

Reply:
I've figured it out, turns out the condition for updating the ballot box was evaluating to false. It's because I was taking the ProveDlog from BallotBoxWrapper::ballot_token_owner, and setting R4 to that. It seems the serialization for that is different than GroupElement which has the type EcPoint

Reply:
> I've figured it out, turns out the condition for updating the ballot box was evaluating to false. It's because I was taking the ProveDlog from BallotBoxWrapper::ballot_token_owner, and setting R4 to that. It seems the serialization for that is different than GroupElement which has the type EcPoint

That's weird. They both hold `EcPoint` and should serialize to the same bytes.

Reply:
Hmm, looking at the source code for Constant/Literal, it seems From<EcPoint> just wraps the EcPoint in the Literal::GroupElement enum, whereas From<ProveDlog> creates a Literal::SigmaProp. I'm guessing their format is different somehow

Reply:
You're right, `ProveDlog` will be `SigmaBoolean` if you make `Literal` out of it. Oh, now I get it. In order to put `ProveDlog` into the register, you need to make a constant and end up with `SigmaBoolean` instead of `GroupElement`. I was thinking of raw serialization only because if you serialize `ProveDlog` it'll just write its `EcPoint` - https://github.com/ergoplatform/sigma-rust/blob/f0caab7d34522284a68cf69b9fab57b58fbfa945/ergotree-ir/src/serialization/data.rs#L42
the same way does `GroupElement` - https://github.com/ergoplatform/sigma-rust/blob/f0caab7d34522284a68cf69b9fab57b58fbfa945/ergotree-ir/src/serialization/data.rs#L42

Should we switch to `EcPoint` from `ProveDlog` in ballot box wrappers to avoid such cases?

Reply:
Hmm, I must admit, I'm not too familliar with the differences between EcPoint and ProveDlog, but there shouldn't be any issue in switching to EcPoint, since the oracle contracts seem to be using EcPoint/GroupElement everywhere too, and there are other instances of having to convert ProveDlog to EcPoint before adding it to a register in ballot_box.rs, extract_reward_tokens, and oracle_box.rs